File: biome.py
from enum import Enum


class Resource(Enum):
    Wheat = 0
    Sheepie = 1
    Lumber = 2
    Ore = 3
    Brick = 4


class Biome:
    primaryRes: Resource = None
    color: str = None

    biomeList: list = []
    '''
    A list of all biomes that are subclassed of this class
    '''

    def __init__(self, *args, **kwargs):
        '''
        Biomes are more like a tag and so shouldn't be
        instantiated.
        '''
        raise TypeError("Biome shouldn't be instantiated")

    def __init_subclass__(cls):
        assert hasattr(cls, 'primaryRes')
        assert isinstance(getattr(cls, 'primaryRes'), Resource)

        assert hasattr(cls, 'color')
        assert isinstance(getattr(cls, 'color'), str)

        Biome.biomeList.append(cls)  # add new biome to list of biomes


class Farmland(Biome):
    primaryRes: Resource = Resource.Wheat
    color = 'goldenrod'


class Grassland(Biome):
    primaryRes: Resource = Resource.Sheepie
    color = 'limegreen'


class Forest(Biome):
    primaryRes: Resource = Resource.Lumber
    color = 'forestgreen'


class Mountain(Biome):
    primaryRes: Resource = Resource.Ore
    color = 'dimgray'


class Quarry(Biome):
    primaryRes: Resource = Resource.Brick
    color = 'firebrick'

File: web/map/Biome.ts
import { defined } from "../util";

export class Biome {
    private color: string;

    protected constructor(color: string) {
        this.color = color;
    }

    public getColor() {
        defined(this.color);
        return this.color;
    }
}

export const Desert = new (class Desert extends Biome {
    constructor() {
        super('yellow');
    }
})();

export const Grassland = new (class Grassland extends Biome {
    constructor() {
        super('limegreen');
    }
})();

export const Forest = new (class Forest extends Biome {
    constructor() {
        super('forestgreen');
    }
})();

export const Mountain = new (class Mountain extends Biome {
    constructor() {
        super('dimgray');
    }
})();

export const Farmland = new (class Farmland extends Biome {
    constructor() {
        super('goldenrod');
    }
})();

export const Quarry = new (class Quarry extends Biome {
    constructor() {
        super('firebrick');
    }
})();

export const biomes: Array<Biome> = [Desert, Grassland, Forest, Mountain, Farmland, Quarry];

export function getBiomeByName(name: string): Biome {
    switch (name) {
        case 'Desert':
            return Desert;
        case 'Grassland':
            return Grassland;
        case 'Forest':
            return Forest;
        case 'Mountain':
            return Mountain;
        case 'Farmland':
            return Farmland;
        case 'Quarry':
            return Quarry;
    
        default:
            throw new Error("KeyError: Could not find biome named '" + name + "'")
    }
}

// export const biomeDistributionArray = function() {
//     let tmp_biomeDistributionArray: Array<Biome> = [];
//     biomes.forEach((e: Biome) => {
//         for (let i = 0; i < e.getAbundance(); i++) {
//             tmp_biomeDistributionArray.push(e);
//         }
//     });
//     return tmp_biomeDistributionArray;
// }();
